realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsManagedTests.java

```diff
@@ -222,8 +234,207 @@ public void onChange(AllJavaTypes object) {
+                    assertEquals(1, counter.get());
+
+                    looperThread.testComplete();
+                }
+            }
+        );
+    }
+
+    // A listener registered on the backlinked field should be called when a commit adds a backlink
+    @Test
+    @RunTestInLooperThread
+    public void notification_onCommitRealmResults() {
+        final Realm looperThreadRealm = looperThread.realm;
+
+        looperThreadRealm.beginTransaction();
+        AllJavaTypes child = looperThreadRealm.createObject(AllJavaTypes.class, 10);
+        looperThreadRealm.commitTransaction();
+
+        final AtomicInteger counter = new AtomicInteger(0);
+        RealmChangeListener<RealmResults<AllJavaTypes>> listener = new RealmChangeListener<RealmResults<AllJavaTypes>>() {
+            @Override
+            public void onChange(RealmResults<AllJavaTypes> object) {
+                counter.incrementAndGet();
+            }
+        };
+        child.getObjectParents().addChangeListener(listener);
+
+        looperThreadRealm.beginTransaction();
+        AllJavaTypes parent = looperThreadRealm.createObject(AllJavaTypes.class, 1);
+        parent.setFieldObject(child);
+        looperThreadRealm.commitTransaction();
+
+        // This should be on the Looper queue, behind the notification
+        looperThread.postRunnable(
+            new Runnable() {
+                @Override
+                public void run() {
+                    assertEquals(2, looperThreadRealm.where(AllJavaTypes.class).findAll().size());
+                    assertEquals(1, counter.get());
+
+                    looperThread.testComplete();
+                }
+            }
+        );
+    }
+
+    // A listener registered on the backlinked field should not be called after the listener is removed
+    @Test
+    @RunTestInLooperThread
+    public void notification_notSentAfterUnregisterListenerRealmResults() {
+        final Realm looperThreadRealm = looperThread.realm;
+
+        looperThreadRealm.beginTransaction();
+        AllJavaTypes child = looperThreadRealm.createObject(AllJavaTypes.class, 10);
+        looperThreadRealm.commitTransaction();
+
+        RealmChangeListener<RealmResults<AllJavaTypes>> listener = new RealmChangeListener<RealmResults<AllJavaTypes>>() {
+            @Override
+            public void onChange(RealmResults<AllJavaTypes> object) {
+                fail("Not expecting notification after unregister");
+            }
+        };
+        RealmResults<AllJavaTypes> objParents = child.getObjectParents();
+        objParents.addChangeListener(listener);
+        objParents.removeChangeListener(listener);
+
+        looperThreadRealm.beginTransaction();
+        AllJavaTypes parent = looperThreadRealm.createObject(AllJavaTypes.class, 1);
+        parent.setFieldObject(child);
+        looperThreadRealm.commitTransaction();
+
+        // This should be on the Looper queue, behind the notification
+        looperThread.postRunnable(
```

Comment:
See comments above. Testing the non-existence of something is not trivial giving the nature of our async notification thread. I'm actually not sure if we have a deterministic way of doing that. The simplest is probably just posting as a delayed runnable with 100ms or something. Alternatively waiting for the onSuccess of an async transaction might also work, but I'm not fully convinced that it is not racy as well.

For simplicity I would probably start by just making this a delayed runnable. Perhaps @beeender has some ideas since he is more familiar with the timing of things at this point.